audio/src/shape_noise.rs

```diff
@@ -0,0 +1,252 @@
+// perceived noise, the higher the absolute power, and the higher the CPU-
+// usage. All shapers have something going for them.
+//
+pub trait NoiseShaper {
+    fn new() -> Self
+    where
+        Self: Sized;
+    fn name(&self) -> &'static str;
+    fn shape(&mut self, sample: f32, noise: f32) -> f32;
+}
+
+impl fmt::Display for dyn NoiseShaper {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.name())
+    }
+}
+
+pub struct NoShaping {}
+impl NoiseShaper for NoShaping {
+    fn new() -> Self {
+        Self {}
+    }
+
+    fn name(&self) -> &'static str {
+        "None"
+    }
+
+    fn shape(&mut self, sample: f32, noise: f32) -> f32 {
+        sample + noise
+    }
+}
+
+// First-order noise shaping. Advantages: negligible performance hit, infinite
+// signal-to-noise ration at DC, lowered signal-to-noise ratio for low
+// frequencies and slightly higher signal-to-noise ration for frequencies
+// above Nyquist/2, kills certain limit cycles in subsequent IIR filters,
+// safe to use on oversampling DACs. Disadvantage: higher perceived loudness
+// than other options (but still less so than without shaping!). Guidance:
+// there are very little reasons not to use this shaper, even on power-
+// constrained hardware. If you do, best to set dithering to none.
+pub struct FractionSaver {
+    active_channel: usize,
+    previous_fractions: [f32; NUM_CHANNELS],
+}
+
+impl NoiseShaper for FractionSaver {
+    fn new() -> Self {
+        Self {
+            active_channel: 0,
+            previous_fractions: [0.0; NUM_CHANNELS],
+        }
+    }
+
+    fn name(&self) -> &'static str {
+        "Fraction Saver"
+    }
+
+    fn shape(&mut self, sample: f32, noise: f32) -> f32 {
+        let sample_with_fraction = sample + noise + self.previous_fractions[self.active_channel];
+        self.previous_fractions[self.active_channel] = sample_with_fraction.fract();
+        self.active_channel = self.active_channel ^ 1;
+        sample_with_fraction.floor()
+    }
+}
+
+// Higher-order noise shapers. Advantage: lower perceived loudness than first-
+// order noise-shaping. Disadvantage: higher CPU-usage, higher absolute noise
+// level, not to be used when there will be dithering or noise shaping further
+// down the chain, as is the case with most delta-sigma DACs. Guidance: use
+// on non-oversampling DACs, or other DACs that do not do dithering or noise
+// shaping themselves.
+macro_rules! fir_shaper {
+    ($name: ident, $description: tt, $taps: expr, $weights: expr) => {
+        pub struct $name {
+            fir: FIR,
+        }
+
+        impl NoiseShaper for $name {
+            fn new() -> Self {
+                Self {
+                    fir: FIR::new(&Self::WEIGHTS),
+                }
+            }
+
+            fn name(&self) -> &'static str {
+                concat!($description, ", ", stringify!(Self::WEIGHTS.len()), " taps")
```

Comment:
Don't think this works. Try this, but I'm not sure:
```suggestion
                concat!($description, ", ", stringify!($taps), " taps")
```